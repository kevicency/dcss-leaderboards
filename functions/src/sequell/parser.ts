import { playerByAccount } from '../data'
import { KilledResult, lgResult, logResult, SequellResult } from './sequell'

export interface ISequellParser {
  parse(message: string): SequellResult
  lgParse(message: string): lgResult
  logParse(message: string): logResult
  killedParse(message: string): KilledResult
}

export class SequellParser implements ISequellParser {
  lgRegex = /^\d+\. \[game_key=([\w:]+)\] (\w+) the ([\w\s]+) \(L(\d+) (\w+)( of ([\w\s]+))?\), escaped with the Orb and (\d+) runes on ([\d\s-:]+), with (\d+) points after (\d+) turns and ([\d:]+)\./
  logRegex = /^\d+\. (\w+), XL(\d+) (\w+), T:(\d+): (.+)$/
  logDateRegex = /.*(\d{4})(\d{2})(\d{2})-(\d{2})(\d{2})(\d{2}).*/
  timoeutRegex = /\d+s limit exceeded: killed (.+)/

  parse(message: string): SequellResult {
    return (
      this.lgParse(message) ||
      this.logParse(message) ||
      this.killedParse(message)
    )
  }

  lgParse(message: string): lgResult {
    if (message) {
      const match = message.match(this.lgRegex)

      if (match && match.length > 1) {
        return {
          type: 'lg',
          gid: match[1],
          account: match[2],
          player: playerByAccount[match[2]] || match[2],
          title: match[3],
          xl: +match[4],
          ...this.parseCharacter(match[5]),
          god: match[7],
          runes: +match[8],
          date: new Date(`${match[9].replace(' ', 'T')}Z`),
          points: +match[10],
          turns: +match[11],
          duration: match[12],
        }
      }
    }
    return null
  }

  logParse(message: string): logResult {
    if (message) {
      const match = message.match(this.logRegex)

      if (match && match.length > 1) {
        return {
          type: 'log',
          account: match[1],
          xl: +match[2],
          ...this.parseCharacter(match[3]),
          turns: +match[4],
          morgue: match[5],
          date: new Date(
            match[5].replace(this.logDateRegex, '$1-$2-$3 $4:$5:$6Z')
          ),
        }
      }
    }
    return null
  }

  killedParse(message: string): KilledResult {
    if (message) {
      const match = message.match(this.timoeutRegex)

      return {
        type: 'killed',
        message: match[1],
      }
    }

    return null
  }

  parseCharacter(character: string) {
    return character
      ? { race: character.slice(0, 2), background: character.slice(2, 4) }
      : { race: null, background: null }
  }
}
