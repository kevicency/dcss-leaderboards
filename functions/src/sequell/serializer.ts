import { uniq } from 'lodash'
import { accounts } from '../data'
import { lgQuery, logQuery, SequellQuery } from './sequell'

export interface ISequellSerializer {
  serialize(query: SequellQuery): string
}

export class SequellSerializer implements ISequellSerializer {
  private ambiguousAcronyms = ['Hu', 'Fe', 'FE']

  serialize(query: SequellQuery) {
    if (query.type === 'lg') {
      return this.lgSerialize(query)
    }

    if (query.type === 'log') {
      return this.logSerialize(query)
    }

    return null
  }

  public logSerialize(query: logQuery) {
    return `!log * gid=${query.gid}`
  }

  public lgSerialize(query: lgQuery) {
    let message = `!lg ${query.player ||
      '*'} win x=gid ${this.characterSerialize(query)}`

    if (query.god) {
      message = `${message} ${query.god}`
    }

    if (query.min) {
      message = `${message} min=${query.min}`
    }

    if (query.playerBlacklist) {
      const accountBlacklist = uniq(
        query.playerBlacklist.reduce(
          (memo, player) => [...memo, player, ...(accounts[player] || [])],
          []
        )
      )

      message = `${message} ${accountBlacklist
        .filter(Boolean)
        .map(account => `!@${account}`)
        .join(' ')}`
    }

    return message.replace(/\s+/g, ' ')
  }

  public characterSerialize({
    race,
    background,
  }: { race?: string; background?: string } = {}) {
    for (const acronym of this.ambiguousAcronyms) {
      if (race === acronym && !background) {
        return `${race}--`
      }
      if (background === acronym && !race) {
        return `--${background}`
      }
    }

    return `${race || ''}${background || ''}`
  }
}
