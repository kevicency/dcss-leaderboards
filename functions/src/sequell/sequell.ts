import { EventEmitter } from 'events'
import * as irc from 'irc'
import { Omit } from 'lodash'
import { createLogger } from '../logger'
import { ISequellParser, SequellParser } from './parser'
import { ISequellSerializer, SequellSerializer } from './serializer'

const logger = createLogger('sequell')

type CharacterQuery = {
  race?: string
  background?: string
  god?: string
}

export type lgQueryFilters = {
  [key: string]: string | number | { op: string; value: string | number }
}
export type lgQuery = CharacterQuery & {
  type: 'lg'
  player?: string
  playerBlacklist?: string[]
  filters?: lgQueryFilters
}

export type logQuery = {
  type: 'log'
  gid: string
}
export type SequellQuery = lgQuery | logQuery

export type lgResult = {
  type: 'lg'
  account: string
  background: string
  date: Date
  duration: string
  gid: string
  god?: string
  player: string
  points: number
  race: string
  title: string
  turns: number
  runes: number
  xl: number
}
export type logResult = {
  type: 'log'
  account: string
  xl: number
  background: string
  race: string
  turns: number
  date: Date
  morgue: string
}
export type KilledResult = {
  type: 'killed'
  message: string
}

export type SequellResult = lgResult | logResult | KilledResult

export interface Sequell extends EventEmitter {
  send(message: string): void
  lg(args: Omit<lgQuery, 'type'>): void
  log(args: Omit<logQuery, 'type'>): void
}

export class IrcSequell extends EventEmitter implements Sequell {
  private irc: irc.Client
  private serializer: ISequellSerializer
  private parser: ISequellParser
  private queue: string[]
  private status: 'initializing' | 'ready' = 'initializing'

  constructor() {
    super()
    this.queue = []
    this.serializer = new SequellSerializer()
    this.parser = new SequellParser()
    this.irc = new irc.Client('irc.freenode.net', 'speedcrawl', {
      port: 6697,
      autoConnect: true,
      autoRejoin: true,
      userName: process.env.IRC_USERNAME,
      password: process.env.IRC_PASSWORD,
      sasl: true,
      secure: true,
      debug: false,
    })
    this.irc.on('message', this.handleIrcMessage)
    this.irc.on('registered', this.handleRegistered)
  }

  private handleIrcMessage = (from: string, to: string, message: string) => {
    logger.debug(`received: ${message}`)

    this.emit('message', from, to, message)

    const result = this.parser.parse(message)

    if (result) {
      this.emit('result', result)
    } else {
      this.emit('unknown-message', from, to, message)
    }
  }

  private handleRegistered = () => {
    this.status = 'ready'

    this.emit('ready')

    if (this.queue.length) {
      do {
        this.send(this.queue.shift())
      } while (this.queue.length)
    }
  }

  public send = (message: string) => {
    if (this.status === 'ready') {
      logger.debug('send: ' + message)

      this.irc.say('sequell', message)
    } else {
      this.queue.push(message)
    }
  }

  lg(query: Omit<lgQuery, 'type'>): void {
    const message = this.serializer.serialize({ type: 'lg', ...query })

    this.send(message)
  }

  log(query: Omit<logQuery, 'type'>): void {
    const message = this.serializer.serialize({ type: 'log', ...query })

    this.send(message)
  }
}
