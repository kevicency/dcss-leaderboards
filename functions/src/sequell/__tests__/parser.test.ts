import 'jest'
import { SequellParser } from '../parser'
import { lgResult } from '../sequell'

const sut = new SequellParser()

test('parse lg', () => {
  const simple =
    '80123. [game_key=idfk:cao:20180603195308S] idfk the Impregnable (L27 DDFi of Makhleb), escaped with the Orb and 3 runes on 2018-07-03 20:21:30, with 1998358 points after 52035 turns and 0:27:40.'

  expect(sut.lgParse(simple)).toMatchObject({
    account: 'idfk',
    background: 'Fi',
    date: new Date('2018-07-03T20:21:30.000Z'),
    duration: '0:27:40',
    gid: 'idfk:cao:20180603195308S',
    god: 'Makhleb',
    player: 'idfk',
    points: 1998358,
    race: 'DD',
    title: 'Impregnable',
    turns: 52035,
    type: 'lg',
    xl: 27,
  } as lgResult)

  const multiWordTitle =
    '80417. [game_key=P0WERM0DE:cue:20180712164931S] P0WERM0DE the Champion of Chaos (L22 DDFi of Makhleb), escaped with the Orb and 3 runes on 2018-08-17 10:01:22, with 3745092 points after 18326 turns and 0:23:13.'
  expect(sut.lgParse(multiWordTitle)).toMatchObject({
    title: 'Champion of Chaos',
  })

  const multiWordGod =
    '1543. [game_key=Sharkman1231:cbro:20180909034635S] Sharkman1231 the Impregnable (L26 DsMo of Nemelex Xobeh), escaped with the Orb and 3 runes on 2018-10-09 06:08:18, with 1738011 points after 63387 turns and 2:07:25.'
  expect(sut.lgParse(multiWordGod)).toMatchObject({
    god: 'Nemelex Xobeh',
  })
})

test('parse log', () => {
  const log =
    '1. pureman, XL21 MiFi, T:24605: https://s3-us-west-2.amazonaws.com/crawl.jorgrun.rocks/morgue/pureman/morgue-pureman-20180807-003530.txt'
  expect(sut.logParse(log)).toMatchObject({
    account: 'pureman',
    background: 'Fi',
    date: new Date('2018-08-07T00:35:30.000Z'),
    morgue:
      'https://s3-us-west-2.amazonaws.com/crawl.jorgrun.rocks/morgue/pureman/morgue-pureman-20180807-003530.txt',
    race: 'Mi',
    turns: 24605,
    type: 'log',
    xl: 21,
  })
})
