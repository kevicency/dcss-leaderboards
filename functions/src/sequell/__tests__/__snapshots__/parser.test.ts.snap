// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse log 1`] = `
Object {
  "account": "pureman",
  "background": "Fi",
  "date": 2018-08-07T00:35:30.000Z,
  "morgue": "https://s3-us-west-2.amazonaws.com/crawl.jorgrun.rocks/morgue/pureman/morgue-pureman-20180807-003530.txt",
  "race": "Mi",
  "turns": 24605,
  "type": "log",
  "xl": 21,
}
`;
